spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Regularizer.scala

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.optim
+
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+import scala.reflect.ClassTag
+
+/**
+ * It is a trait for all regularizers.
+ * Any regularizers need to inherit the result.
+ *
+ * @tparam T type parameters [[Float]] or [[Double]]
+ */
+trait Regularizer[T]
+  extends Serializable {
+  var isRegualrized: Boolean = true
```

Comment:
accRegularization is called inside layers, so the container may not be able to control whether to do accRegularization outside directly, the isRegualrized has to be passed into Regularizer.

`isRegualrized` will be set by `Cell` class, so I do not think it should be protected, it need to be called by some classes outside the `com.intel.analytics.bigdl.optim` package.